GetStream/stream-chat-android
# [3814] cancel children jobs when ChatClient gets disconnected

Related PR: https://github.com/GetStream/stream-chat-android/pull/3960

### 🎯 Goal

Cancel all jobs attached to SDK scope except the `disconnect` job.


### 🧪 Testing

- Apply the patch below
- Install any sample app
- Login 
- Interact with the app
- Logout
- Observe logcat

Expected result:
```
... [disconnectSuspend] >>> before cancellation: 10
... [disconnectSuspend] <<< after  cancellation: 1
```

<details>

<summary>Provide the patch summary here</summary>

```
Index: stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt b/stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt
--- a/stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt	(revision 87a81af810078178b97af22935148753bf034022)
+++ b/stream-chat-android-client/src/main/java/io/getstream/chat/android/client/ChatClient.kt	(date 1660302259227)
@@ -1095,8 +1095,10 @@
         lifecycleObserver.dispose()
         appSettingsManager.clear()
         _repositoryFacade = null
+        logger.i { "[disconnectSuspend] >>> before cancellation: ${scope.coroutineContext[Job]?.children?.count()}" }
         val currentJob = currentCoroutineContext()[Job]
         scope.coroutineContext.cancelChildrenExcept(currentJob)
+        logger.i { "[disconnectSuspend] <<< after  cancellation: ${scope.coroutineContext[Job]?.children?.count()}" }
     }
 
     /**
```

</details>


### ☑️Contributor Checklist

#### General

#### Code & documentation

### ☑️Reviewer Checklist

### 🎉 GIF

_Please provide a suitable gif that describes your work on this pull request_
